### PR TITLE
Added basic featurizers and their test cases

### PR DESCRIPTION
**Pull request by rthapa84**

### CODE REVIEW

File: src/piton/featurizers/featurizers_notes.py [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+
+class NoteFeaturizer:
+    def __init__(
+        self,
+        path_to_patient_database: str = "",
+        path_to_tokenizer: str = "",
+        path_to_embedder: str = "",
+        path_to_temp_dir: str = "",
+        path_to_output_dir: str = "",
+        preprocess_transformations: List = [],
```
Comment (Miking98) [inline src/piton/featurizers/featurizers_notes.py:151]: Type: `List[Callable]`

File: src/piton/featurizers/featurizers_notes.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+        notes_for_labels: List[
+            PatientLabelNotesTuple
+        ] = load_preprocessed_note_chunk(path_to_temp_dir, chunk_id)
+
+        # Parse arguments
+        tokenizer = AutoTokenizer.from_pretrained(
+            path_to_tokenizer, use_fast=True, batched=True
+        )
+        max_length = params.get("tokenizer_max_length", 512)
```
Comment (Miking98) [inline src/piton/featurizers/featurizers_notes.py:286]: Change default to `None`

File: src/piton/featurizers/featurizers_notes.py [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+            == np_label_idxs.shape[0]
+            == embeddings.shape[0]
+        )
+        result_tuple: Tuple[NDArray, NDArray, NDArray] = (
+            embeddings,
+            np_patient_ids,
+            np_label_idxs,
+        )
+        save_to_pkl(
+            result_tuple, os.path.join(self.path_to_output_dir, "features.pkl")
```
Comment (Miking98) [inline src/piton/featurizers/featurizers_notes.py:557]: Rename to be consistent with how other featurizers save / return their results

File: .gitignore [hunk 4]
```diff
@@ -3,4 +3,4 @@ __pycache__
 *.so
 bazel-*
 *.egg-info
-tutorials/*
+# tutorials/*
```
Comment (Lalaland) [inline .gitignore:6]: This seems like a bad idea

**Comment by Lalaland** [comment]
@rthapa84 I don't think it makes sense for me to review this until you clean it up a bit more? There is a lot of junk changes floating around in this PR, in particular in the native folder, that don't seem intentional?

Can you double check that all the changed files are intentional?

File: src/piton/featurizers/core.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+        i.e. the featurizer doesn't need preprocessing.
+
+        Args:
+            patient (Patient): A patient to preprocess on.
+            labels (List[Label]): The list of labels of this patient to preprocess on.
+        """
+        pass
+
+    @classmethod
+    def aggregate_featurizers(
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:205]: I don't think a default implementation makes sense for this function. Should be abstract.

File: src/piton/featurizers/core.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        patient: Patient,
+        labels: List[Label],
+        ontology: Optional[Ontology],
+    ) -> List[List[ColumnValue]]:
+        """Featurize the patient such that each label in `labels` has an associated list of features.
+
+        Example:
+            return [
+                [ ColumnValue(0, 10), ColumnValue(3, 12), ], # features for label 0
+                [ ColumnValue(1, 'hi') ], # features for label 1
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:239]: This example with 'hi' is wrong. Features must be numeric.

File: src/piton/featurizers/core.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                [ ColumnValue(2, 2), ColumnValue(1, 3), ColumnValue(10, 3), ], # features for label 2
+                ...
+                [ ColumnValue(8, True), ColumnValue(9, False), ], # features for label n
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:242]: Likewise, True and False are wrong, must be numeric.

File: src/piton/featurizers/core.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+        """
+        return "no name"
+
+    def is_needs_preprocessing(self) -> bool:
+        """Return TRUE if you must run `preprocess()`. If FALSE, then `preprocess()`
+        should do nothing.
+        """
+        return False
+
+    def to_dict(self) -> Dict[str, Any]:
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:279]: Remove to and from dict as we should be using pickle.

File: src/piton/featurizers/core.py [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+
+    def __init__(self, featurizers: List[Featurizer]):
+        """Create a `FeaturizerList` from a sequence of featurizers.
+
+        Args:
+            featurizers (List[Featurizer]): The featurizers to use for featurizing patients.
+        """
+        self.featurizers: List[Featurizer] = featurizers
+
+    # TODO - restructure code so that multiprocessing happens within the featurizer
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:307]: Remove this TODO. The current setup is actually better.

File: src/piton/featurizers/core.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+            trained_featurizers: List[Featurizer] = [
+                y
+                for x in pool.imap(_run_preprocess_featurizers, tasks)
+                for y in x
+            ]
+        grouped_featurizers: collections.defaultdict = collections.defaultdict(
+            list
+        )
+        for featurizer in trained_featurizers:
+            featurizer_name = featurizer.__class__.__name__
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:345]: Why do you need featurizer names here? Why not just use indices from the original list? Featurizer 0, featurizer 1, etc, etc.

File: src/piton/featurizers/core.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+            results: List[Tuple[Any, Any, Any, Any]] = list(
+                pool.imap(_run_featurizer, tasks)
+            )
+
+        # Join results
+        data_matrix = scipy.sparse.vstack([x[0] for x in results])
+        label_pids: NDArray[Literal["n_labels, 1"], np.int64] = np.concatenate(
+            [x[1] for x in results]
+        )
+        label_values: NDArray[Literal["n_labels, 1"], Any] = np.concatenate(
```
Comment (Lalaland) [inline src/piton/featurizers/core.py:399]: Not sure this code would work for SurvivalValues? Does np.concatenate work with arbitrary Python objects?

File: src/piton/featurizers/featurizers.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+from .. import Patient
+from ..extension import datasets as extension_datasets
+from ..labelers.core import Label
+from . import Dictionary, OnlineStatistics
+from .core import ColumnValue, Featurizer
+
+
+# TODO - replace this with a more flexible/less hacky way to allow the user to
+# manage patient attributes (like age)
+def get_patient_birthdate(patient: Patient) -> Optional[datetime.datetime]:
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers.py:17]: This should never return None.

Throw an exception if there is no birth date.

File: src/piton/featurizers/featurizers.py [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+            if self.is_normalize:
+                # age = (age - mean(ages)) / std(ages)
+                age_in_yrs = (age_in_yrs - self.age_statistics.mean()) / (
+                    self.age_statistics.standard_deviation()
+                )
+            all_columns.append([ColumnValue(0, age_in_yrs)])
+
+        return all_columns
+
+    def to_dict(self) -> Dict[str, Any]:
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers.py:98]: Replace to and from dict with pickle

File: src/piton/featurizers/featurizers.py [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+                                codes_per_bin[i].popleft()
+                            )
+
+                            code_counts_per_bin[i][next_code] -= 1
+                            if code_counts_per_bin[i][next_code] == 0:
+                                del code_counts_per_bin[i][next_code]
+
+                            code_counts_per_bin[i + 1][next_code] += 1
+
+                # print(codes_per_bin, " | ", code_counts_per_bin)
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers.py:288]: We should remove these commented out print statements

File: src/piton/transforms/notes.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    text: str = " ".join([note[1].value for note in notes])  # type: ignore
+    # Give it an arbitrary `start` and `code` (b/c merged notes don't have one)
+    note = Event(start=datetime.datetime(0, 0, 0), code=0, value=text)
+    return [(0, note)]
+
+
+def keep_only_last_n_chars(
+    notes: NotesProcessed, label: Label, **kwargs
+) -> NotesProcessed:
+    """Keep the last `n_chars` from each note."""
```
[the remark this answers is not in the document]
Comment (Lalaland) [reply src/piton/transforms/notes.py:66]: Also, when would you ever use this transformation? Seems pointless?

File: tests/tools.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                    "valid_end_DATE": "2099-12-31",
+                    "invalid_reason": "",
+                    "load_table_id": "custom_mapping",
+                    "load_row_id": "",
+                }
+            )
+    return concept_map
+
+
+class DummyOntology:
```
Comment (Lalaland) [inline tests/tools.py:129]: Why are we creating a dummy ontology class here? If you are already going through the trouble of creating a real PatientDatabase, why not use the associated real ontology for that database?

**Review by Lalaland — CHANGES_REQUESTED** [review]
Initial first pass of a review. Haven't looked deeply over the text stuff or the files that Michael noted would be ovewritten.

File: pyproject.toml [hunk 17]
```diff
@@ -11,6 +11,10 @@ dependencies = [
     "scikit-learn >= 0.24",
     "tqdm >= 4.60.0",
     "zstandard >= 0.18",
+    "nptyping == 2.4.1",
+    "torchtyping == 0.1.4",
+    "transformers == 4.25.1",
```
[the remark this answers is not in the document]
Comment (Lalaland) [reply pyproject.toml:16]: Oh, we need it for the text tokenization. We might want to move that to an optional dependency. That's a very heavy dependency to add.

File: src/piton/featurizers/featurizers_notes.py [hunk 18]
```diff
@@ -0,0 +1,570 @@
...
+
+def print_log(name: str, content: str):
+    print(f"{int(time.time() - START_TIME)} | {name} | {content}")
+
+
+def save_to_pkl(object_to_save, path_to_file: str):
+    """Save object to Pickle file."""
+    os.makedirs(os.path.dirname(path_to_file), exist_ok=True)
+    with open(path_to_file, "wb") as fd:
+        print(fd is None, object_to_save is None)
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers_notes.py:54]: Having random print statements like this throughout the codebase tends to be suboptimal

File: src/piton/featurizers/featurizers_notes.py [hunk 19]
```diff
@@ -0,0 +1,570 @@
...
+                model = model.to(device)
+            except Exception as e:
+                print_log("embed", str(e))
+                raise InterruptedError(
+                    f"Error loading model onto device {device}"
+                )
+            print_log("embed", f"chunk #{chunk_id} | device {device}")
+
+            # Setup dataloader for model
+            batch_size: int = params.get("batch_size", 4096)
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers_notes.py:369]: Seems like a rather large batch size?

File: src/piton/featurizers/featurizers_notes.py [hunk 20]
```diff
@@ -0,0 +1,570 @@
...
+                token_embeddings.shape[1]
+                == notes_tokenized["input_ids"].shape[1]
+            )
+
+            # Derive a single embedding representation for this note from its token embeddings
+            # Note: Need clone(), otherwise will keep original Tensor when save to file
+            #   see: https://discuss.pytorch.org/t/saving-tensor-with-torch-save-uses-too-much-memory/46865
+            #   or: https://pytorch.org/docs/stable/notes/serialization.html#preserve-storage-sharing
+            print_log("embed", f"chunk #{chunk_id} | aggregating embeddings")
+            note_embeddings: NotesEmbedded = embed_method(
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers_notes.py:424]: Shouldn't these be serialized as numpy to make them easier to load later?

File: src/piton/featurizers/featurizers_notes.py (hunk 1 above)
Comment (rthapa84) [reply src/piton/featurizers/featurizers_notes.py:163]: Fixed.

File: pyproject.toml (hunk 17 above)
[the remark this answers is not in the document]
Comment (rthapa84) [reply pyproject.toml:16]: Moved to optional.

File: src/piton/__init__.py [hunk 21]
```diff
@@ -79,3 +77,12 @@ def __eq__(self, other: object) -> bool:
     def __repr__(self) -> str:
         val_str = ", ".join(f"{a}={b}" for a, b in self.__dict__.items())
         return f"Event({val_str})"
+
+    def __getstate__(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (rthapa84) [reply src/piton/__init__.py:81]: Not sure who made this change. @Miking98 ?

File: src/piton/__init__.py [hunk 22]
```diff
@@ -55,15 +55,13 @@ def __init__(
...
-                self.__dict__[a] = b
+            self.__dict__[a] = b
 
     def __getattr__(self, __name: str) -> Any:
         return None
 
     def __setattr__(self, name: str, value: Any) -> None:
-        if value is not None:
-            self.__dict__[name] = value
+        self.__dict__[name] = value
```
[the remark this answers is not in the document]
Comment (rthapa84) [reply src/piton/__init__.py:64]: Not sure who made this changes. @Miking98 ?

File: src/piton/featurizers/core.py (hunk 8 above)
Comment (rthapa84) [reply src/piton/featurizers/core.py]: @Miking98 Fixing it.

File: tests/tools.py (hunk 16 above)
Comment (rthapa84) [reply tests/tools.py:129]: removed ontology class. All we need is a list of concept really.

File: src/piton/featurizers/core.py [hunk 23]
```diff
@@ -0,0 +1,420 @@
...
+        np_data.shape == np_indices.shape
+    ), f"`data` should have equal shape as `indices`, but instead have {np_data.shape} != {np_indices.shape}"
+    data_matrix = scipy.sparse.csr_matrix(
+        (np_data, np_indices, np_indptr), shape=(total_rows, total_columns)
+    )
+
+    label_pids: NDArray[Literal["n_labels, 1"], np.int64] = np.array(
+        [x[0] for x in label_data], dtype=np.int64
+    )
+    label_values: NDArray[Literal["n_labels, 1"], Any] = np.array(
```
[the remark this answers is not in the document]
Comment (rthapa84) [reply src/piton/featurizers/core.py:148]: Not sure how we should go about testing this?

File: src/piton/featurizers/featurizers_notes.py [hunk 24]
```diff
@@ -0,0 +1,570 @@
...
+    def get_num_columns(self) -> int:
+        return 768
+
+    def is_needs_preprocessing(self) -> bool:
+        return False
+
+    def get_name(self) -> str:
+        return "TextFeaturizer"
+
+    def preprocess_parallel(self, args: Tuple) -> List[PatientLabelNotesTuple]:
```
[the remark this answers is not in the document]
Comment (rthapa84) [reply src/piton/featurizers/featurizers_notes.py:202]: Sorry, which method are you talking about? If it is the get_name, that has been removed.

[the remark this answers is not in the document]
Comment (Lalaland) [reply src/piton/featurizers/featurizers_notes.py:210]: This was a comment on preprocess_parallel.

File: src/piton/featurizers/core.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (Lalaland) [reply src/piton/featurizers/core.py:148]: By writing some test cases in the test_labeler.py file

File: src/piton/featurizers/__init__.py [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    @classmethod
+    def merge(cls, stats_list: List[OnlineStatistics]) -> OnlineStatistics:
+        """
+        Merge a list of online statistics.
+        """
+        if len(stats_list) == 0:
+            raise ValueError("Cannot merge an empty list of statistics.")
+        unmerged_stats = copy.deepcopy(stats_list)
+        stats1: OnlineStatistics = unmerged_stats.pop(0)
+        while len(unmerged_stats) > 0:
```
Comment (Lalaland) [inline src/piton/featurizers/__init__.py:112]: You are supposed to run the merging algorithm in a tree like structure to get it to perform optimally.

Given the array [a, b, c, d]

ab = merge(a, b)
cd = merge(c, d)

result = merge(ab, cd)

Should be a pretty simple change to the code and will give a lot better performance.

Also, we really should have unit tests for this class.

File: src/piton/featurizers/featurizers.py [hunk 26]
```diff
@@ -0,0 +1,435 @@
...
+
+            code_counts_per_bin: Dict[int, Dict[int, int]] = {
+                i: defaultdict(int) for i in range(len(self.time_bins) + 1)
+            }
+
+            label_idx = 0
+            for event in patient.events:
+                code: int = event.code  # type: ignore
+                while event.start > labels[label_idx].time:
+                    label_idx += 1
```
Comment (Lalaland) [inline src/piton/featurizers/featurizers.py:346]: Whenever you increment label_idx, don't you have to shuffle the code counter per bin?

Comment (rthapa84) [reply src/piton/featurizers/featurizers.py:394]: Done.

File: src/piton/labelers/core.py [hunk 27]
```diff
@@ -389,3 +448,20 @@ def get_patient_start_end_times(
     def get_labeler_type(self) -> LabelType:
         """Return boolean labels (TRUE if event occurs in TimeHorizon, FALSE otherwise)."""
         return "boolean"
+
+
+class OneLabelPerPatient(LabelingFunction):
```
[the remark this answers is not in the document]
Comment (rthapa84) [reply src/piton/labelers/core.py]: Done.

**Review by Lalaland — APPROVED** [review]

**Pull request merged**